### PR TITLE
Refactor bundle-remove to handle multiple bundles

### LINKED ISSUE #449: Refactor bundle-remove to not call remove_bundle() multiple times
The current `remove_bundle()` function was designed to only be called once, but for multiple bundle removals, it is currently called multiple times.

An improvement would be to refactor the bundle removal code to handle multiple bundles at once.

### PR DESCRIPTION
**Pull request by castulo**
The current remove_bundle() function was designed to only be called
once, but for multiple bundle removals, it is currently called
multiple times.

This commit refactors this function so the bundle removal code can
handle multiple bundles at once.

Fixes #449

### CODE REVIEW

File: src/bundle.c [hunk 1]
```diff
@@ -411,10 +411,13 @@ int show_bundle_reqd_by(const char *bundle_name, bool server)
  *  	performing a unlink(2) for each filename.
  *  6) Done.
  */
-int remove_bundle(const char *bundle_name)
+int remove_bundles(char **bundles)
```
Comment (otaviobp) [inline src/bundle.c]: Swupd in general isn't strict about having const in variables that aren't changed in the function, but as this function was already adding a const, I would keep that.

Comment (otaviobp) [reply src/bundle.c]: Also update the header documentation saying that the function now handle multiple bundles

File: src/bundle.c [hunk 2]
```diff
@@ -426,126 +429,157 @@ int remove_bundle(const char *bundle_name)
...
+			  current_version,
+			  ret);
+		if (ret) {
+			fprintf(stderr, "Error: Bundle remove failed\n");
+			retCode = ret;
+		}
+	}
+
+	if (bad > 0) {
+		fprintf(stderr, "%i bundle(s) of %i failed to remove\n", bad, total);
```
Comment (otaviobp) [inline src/bundle.c]: I know you are not adding this line, but it would be nice to correct this message (in another commit) to something like: "%i bundle(s) of %i failed to be removed".
And to be honest I don't even know if my suggestion is correct.

File: test/functional/bundleremove/remove-multiple/test.bats [hunk 3]
```diff
@@ -32,9 +32,19 @@ teardown() {
 }
 
 @test "bundle-remove remove multiple bundles one exist and one that dont" {
-  run sudo sh -c "$SWUPD bundle-remove $SWUPD_OPTS test-bundle3 fake-bundle"
+  run sudo sh -c "$SWUPD bundle-remove $SWUPD_OPTS fake-bundle test-bundle3"
 
-  [ ! -f "$DIR/target-dir/test-file3" ] && [ $status -eq 3 ]
+  # EBUNDLE_NOT_TRACKED == 13
```
Comment (otaviobp) [inline test/functional/bundleremove/remove-multiple/test.bats]: +1 for adding more tests

File: test/functional/bundleremove/remove-multiple/test.bats [hunk 4]
```diff
@@ -32,9 +32,19 @@ teardown() {
 }
 
 @test "bundle-remove remove multiple bundles one exist and one that dont" {
-  run sudo sh -c "$SWUPD bundle-remove $SWUPD_OPTS test-bundle3 fake-bundle"
+  run sudo sh -c "$SWUPD bundle-remove $SWUPD_OPTS fake-bundle test-bundle3"
```
Comment (otaviobp) [inline test/functional/bundleremove/remove-multiple/test.bats]: why did you change the order of the bundles on the test?

File: src/bundle.c (hunk 1 above)
Comment (castulo) [reply src/bundle.c]: Good point.

File: test/functional/bundleremove/remove-multiple/test.bats (hunk 3 above)
Comment (castulo) [reply test/functional/bundleremove/remove-multiple/test.bats]: ;)

File: test/functional/bundleremove/remove-multiple/test.bats (hunk 4 above)
Comment (castulo) [reply test/functional/bundleremove/remove-multiple/test.bats]: well, because I wanted to make sure the exit code of the bundle that failed to be removed was indeed being kept to be returned once all bundles had finished being removed. 
So I changed the order so the test would have a "bad remove" followed by a "good remove", but the exit code from the bad remove should not get lost.

**Comment by castulo** [comment]
Thanks for your review @otaviobp

File: test/functional/bundleremove/remove-multiple/test.bats (hunk 4 above)
Comment (otaviobp) [reply test/functional/bundleremove/remove-multiple/test.bats]: Got it. It makes sense.

File: src/clr_bundle_rm.c [hunk 5]
```diff
@@ -191,84 +191,11 @@ static bool parse_options(int argc, char **argv)
 	return false;
 }
 
-static void reload_parsed_opts(void)
-{
-	if (curopts.path_prefix) {
```
Comment (otaviobp) [inline src/clr_bundle_rm.c]: I think you can get rid of the curopts struct as you are not using the reload_parsed_opts() anymore. can you check that?

File: src/bundle.c (hunk 1 above)
Comment (castulo) [reply src/bundle.c]: Turns out the compiler didn't like that, after adding the const I got a compiler warning:

src/clr_bundle_rm.c: In function ‘bundle_remove_main’:
src/clr_bundle_rm.c:200:24: warning: passing argument 1 of ‘remove_bundles’ from incompatible pointer type [-Wincompatible-pointer-types]
  return remove_bundles(bundles);
                        ^~~~~~~
In file included from src/clr_bundle_rm.c:35:
src/swupd.h:380:40: note: expected ‘const char **’ but argument is of type ‘char **’
 extern int remove_bundles(const char **bundles);
                           ~~~~~~~~~~~~~^~~~~~~

File: src/clr_bundle_rm.c (hunk 5 above)
Comment (castulo) [reply src/clr_bundle_rm.c]: good catch, I'll remove that in another commit.

File: src/bundle.c [hunk 6]
```diff
@@ -426,126 +433,157 @@ int remove_bundle(const char *bundle_name)
...
+			bad++;
+			goto out_free_mom;
+		}
 
-	/* deduplication needs file list sorted by filename, do so */
-	bundle_manifest->files = list_sort(bundle_manifest->files, file_sort_filename);
-	deduplicate_files_from_manifest(&bundle_manifest, current_mom);
+		/* check if bundle is required by another installed bundle */
+		struct list *reqd_by = NULL;
+		required_by(&reqd_by, *bundles, current_mom, 0);
```
Comment (icarus-sparry) [inline src/bundle.c]: How on earth did I let required_by get into the code base - I must have been asleep!

File: src/bundle.c [hunk 7]
```diff
@@ -426,126 +433,157 @@ int remove_bundle(const char *bundle_name)
...
-		goto out_free_curl;
-	}
+	for (; *bundles; ++bundles, total++) {
 
-	if (!search_bundle_in_manifest(current_mom, bundle_name)) {
-		fprintf(stderr, "Bundle name is invalid, aborting removal\n");
-		ret = EBUNDLE_REMOVE;
-		goto out_free_mom;
-	}
+		fprintf(stderr, "Removing bundle: %s\n", *bundles);
```
Comment (icarus-sparry) [inline src/bundle.c]: This print ought to be after the check it is not os-core.

File: src/bundle.c [hunk 8]
```diff
@@ -426,126 +433,157 @@ int remove_bundle(const char *bundle_name)
...
 
 	mix_exists = (check_mix_exists() & system_on_mix());
 
-	current_mom = load_mom(current_version, false, mix_exists);
-	if (!current_mom) {
-		fprintf(stderr, "Unable to download/verify %d Manifest.MoM\n", current_version);
-		ret = EMOM_NOTFOUND;
-		goto out_free_curl;
-	}
+	for (; *bundles; ++bundles, total++) {
```
Comment (icarus-sparry) [inline src/bundle.c]: Not a fan of 130 line for loops, hard to review on github!

I would create a new variable to avoid the need to dereference *bundles all the time.

File: src/bundle.c [hunk 9]
```diff
@@ -399,7 +399,11 @@ int show_bundle_reqd_by(const char *bundle_name, bool server)
 /*  This function is a fresh new implementation for a bundle
  *  remove without being tied to verify loop, this means
  *  improved speed and space as well as more roubustness and
- *  flexibility. What it does is basically:
+ *  flexibility. The function removes one ore more bundles
```
[the remark this answers is not in the document]
Comment (castulo) [reply src/bundle.c]: Good catch, I'll update that.

File: src/bundle.c (hunk 7 above)
Comment (castulo) [reply src/bundle.c]: Ok

File: src/bundle.c (hunk 8 above)
Comment (castulo) [reply src/bundle.c]: Done, I'll create a new variable to store the bundle in the next commit.

File: src/bundle.c (hunk 1 above)
Comment (matthewrsj) [reply src/bundle.c]: The compiler didn't like it because you removed `const` from `src/swupd.h`

Comment (matthewrsj) [reply src/bundle.c]: /me reads you error message closer.. nvm

File: src/bundle.c [hunk 10]
```diff
@@ -411,10 +415,13 @@ int show_bundle_reqd_by(const char *bundle_name, bool server)
  *  	performing a unlink(2) for each filename.
  *  6) Done.
  */
-int remove_bundle(const char *bundle_name)
+int remove_bundles(char **bundles)
 {
 	int lock_fd;
 	int ret = 0;
+	int retCode = 0;
```
Comment (matthewrsj) [inline src/bundle.c]: camel case please

**Comment by matthewrsj** [comment]
@icarus-sparry and @otaviobp any further comments?

File: src/bundle.c (hunk 10 above)
Comment (castulo) [reply src/bundle.c]: ups, my bad! Changing it.
